pyuoi/linear_model/base.py

```diff
@@ -0,0 +1,505 @@
+        # set up data arrays
+        estimates = np.zeros(               # coef_ for each bootstrap for each support
+            (
+                self.n_boots_est,
+                self.n_selection_thresholds * self.n_reg_params_,
+                self.n_features
+            ),
+            dtype=np.float32
+        )
+        self.scores_ = np.zeros(            # score (r2/AIC/AICc/BIC) for each bootstrap for each support
+            (
+                self.n_boots_est,
+                self.n_selection_thresholds * self.n_reg_params_
+            ),
+            dtype=np.float32
+        )
+
+        # iterate over bootstrap samples
+        for bootstrap in trange(
+            self.n_boots_est, desc='Model Estimation', disable=not verbose
+        ):
+
+            # draw a resampled bootstrap
+            X_train, X_test, y_train, y_test = train_test_split(
+                X, y,
+                train_size=self.selection_frac,
+                stratify=stratify,
+                random_state=self.random_state
+            )
+            X_rep, y_rep = resample(X, y)
+
+            # iterate over the regularization parameters
+            for rp_idx, lamb in enumerate(self.reg_params_):
+                # extract current support set
+                support = self.supports_[rp_idx]
+                # if nothing was selected, we won't bother running OLS
+                if np.any(support):
+                    # compute ols estimate
+                    self.estimation_lm.fit(
+                        X_train[:, support],
+                        y_train
+                    )
+
+                    # store the fitted coefficients
+                    estimates[bootstrap, rp_idx, support] = self.estimation_lm.coef_.ravel()
+
+                    # obtain predictions for scoring
+                    y_pred = self.estimation_lm.predict(X_test[:, support])
+                else:
+                    # no prediction since nothing was selected
+                    y_pred = np.zeros(y_test.size)
+
+
+                # calculate estimation score
+                self.scores_[bootstrap, lamb_idx] = self.score_predictions(
+                    score=self.estimation_score,
+                    y_true=y_test,
+                    y_pred=y_pred,
+                    n_features=np.count_nonzero(support),
+                    metric=self.estimation_score,
+                )
+
+        self.lambda_max_idx = np.argmax(self.scores_, axis=1)
+        # extract the estimates over bootstraps from model with best lambda
+        best_estimates = estimates[
+            np.arange(self.n_boots_est), self.lambda_max_idx, :
+        ]
+        # take the median across estimates for the final, bagged estimate
+        self.coef_ = np.median(best_estimates, axis=0)
+
+        self._set_intercept(X_offset, y_offset, X_scale)
+
+        return self
+
+    @staticmethod
+    def _stability_selection_to_threshold(stability_selection, n_boots):
```

Comment:
Is this only used in linear models, or should be it somewhere else like in a utils.py?

Reply:
This has no utility in the dimensionality reduction scenario, so it seems specific to linear models. I'm open to moving it into a linear model specific utils module.